packages/devtools_app/lib/src/memory/flutter/memory_graph_model.dart

```diff
@@ -0,0 +1,439 @@
+    if (classNamesToMonitor != null &&
+        classNamesToMonitor.isNotEmpty &&
+        classNamesToMonitor.contains(className)) {
+      print('WARNING: class $className is monitored.');
+      _monitorClasses.putIfAbsent(i, () => className);
+    }
+
+    classes[i] = HeapGraphClassActual(c);
+  }
+
+  // Pre-allocate the number of objects in the snapshot.
+  final List<HeapGraphElementActual> elements =
+      List<HeapGraphElementActual>(graph.objects.length);
+
+  // Construct all objects.
+  for (int i = 0; i < graph.objects.length; i++) {
+    final HeapSnapshotObject o = graph.objects[i];
+    elements[i] = HeapGraphElementActual(o);
+  }
+
+  // Associate each object with a Class.
+  for (int i = 0; i < graph.objects.length; i++) {
+    final HeapSnapshotObject o = graph.objects[i];
+    final HeapGraphElementActual converted = elements[i];
+    if (o.classId == 0) {
+      // classId of zero is a sentinel.
+      converted.theClass = classSentinel;
+    } else {
+      // Allows finding and debugging a class in the snapshot.
+      // classIds in the object are 1 based need to make zero based.
+      monitorClass(classId: o.classId - 1);
+
+      converted.theClass = classes[o.classId - 1];
+    }
+
+    // Lazily compute the references.
+    converted.referencesFiller = () {
+      for (int refId in o.references) {
+        HeapGraphElement ref;
+        if (refId == 0) {
+          ref = elementSentinel;
+        } else {
+          ref = elements[refId - 1];
+        }
+        converted.references.add(ref);
+      }
+    };
+  }
+
+  return HeapGraph(
+    controller,
+    builtInClasses,
+    classSentinel,
+    classes,
+    elementSentinel,
+    elements,
+  );
+}
+
+class HeapGraph {
+  HeapGraph(
+    this.controller,
+    this.builtInClasses,
+    this.classSentinel,
+    this.classes,
+    this.elementSentinel,
+    this.elements,
+  );
+
+  final MemoryController controller;
+
+  bool instancesComputed = false;
+
+  /// Known built-in classes.
+  final Map<String, int> builtInClasses;
+
+  /// Sentinel Class, all class sentinels point to this object.
+  final HeapGraphClassSentinel classSentinel;
+
+  /// Indexed by classId.
+  final List<HeapGraphClassActual> classes;
+
+  /// Sentinel Object, all object sentinels point to this object.
+  final HeapGraphElementSentinel elementSentinel;
+
+  /// Index by objectId.
+  final List<HeapGraphElementActual> elements;
+
+  /// Group all classes by all libraries.
+  final Map<String, List<HeapGraphClassActual>> rawGroupByLibrary = {};
+
+  /// Group all classes by libraries - with applied filters.
+  final Map<String, List<HeapGraphClassActual>> groupByLibrary = {};
+
+  /// Group all instances by all classes.
+  final Map<String, List<HeapGraphElementActual>> rawGroupByClass = {};
+
+  /// Group all instances by all classes - with applied filters.
+  final Map<String, List<HeapGraphElementActual>> groupByClass = {};
+
+  String normalizeLibraryName(HeapSnapshotClass theClass) {
```

Comment:
document what the normalization is. Also move to a shared location. we'll want to normalize with other normalization schemes used in the debugger and in the future in the inspector.

Reply:
Done.